lib/parser/provenance.h

```diff
@@ -83,21 +86,47 @@ class Provenance {
 
 using ProvenanceRange = common::Interval<Provenance>;
 
+// Maps contiguous ranges of byte offsets in original source files to
+// contiguous ranges in the cooked character stream; essentially a
+// partial inversion of OffsetToProvenanceMappings (below).
+// Used for implementing the first step of mapping an identifier
+// selected in a code editor to one of its declarative statements.
+class ProvenanceRangeToOffsetMappings {
+public:
+  ProvenanceRangeToOffsetMappings();
+  ~ProvenanceRangeToOffsetMappings();
+  bool empty() const { return map_.empty(); }
+  void Put(ProvenanceRange, std::size_t offset);
+  std::optional<std::size_t> Map(ProvenanceRange) const;
+  std::ostream &Dump(std::ostream &) const;
+
+private:
+  // A comparison function object for use in std::multimap<Compare=>.
+  // Intersecting intervals will effectively compare equal, not being
+  // either < nor >= each other.
+  struct WhollyPrecedes {
+    bool operator()(ProvenanceRange, ProvenanceRange) const;
+  };
+
+  std::multimap<ProvenanceRange, std::size_t, WhollyPrecedes> map_;
```

Comment:
Does this have to be multimap?

Reply:
Perhaps not now, after I've given up on mapping macro definitions.  But it probably doesn't matter either way, all probes use `equal_range()`.